### PR TITLE
[ISSUE-4268] Changes for database support

### PR DESCRIPTION
**Pull request by leandroBorgesFerreira**
### 🎯 Goal

Adding changes to SDK to make the process to change the database possible. 

### 🛠 Implementation details
- Adding `StreamOfflinePluginFactory.setCustomRepositoryFactory` to add a different repository to the offline module. 
- Adding `QuerySortParser` as a public class to create a `QuerySorter` out of saved sort specifications.
- Make `SortSpecification` public so it is possible to save it in the database.

### 🧪 Testing

The test can be done in [this PR](https://github.com/GetStream/stream-chat-android/pull/4323): 

#### General
- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] Assigned a person / code owner group (required)
- [x] Thread with the PR link started in a respective Slack channel (#android-chat-core or #android-chat-ui) (required)
- [x] PR targets the `develop` branch
- [x] PR is linked to the GitHub issue it resolves

#### Code & documentation
- [ ] Changelog is updated with client-facing changes
- [x] New code is covered by unit tests
- ~[ ] Comparison screenshots added for visual changes~
- [x] Affected documentation updated (KDocs, docusaurus, tutorial)

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] Compose sample runs & works
- [ ] UI Changes correct (before & after images)
- [ ] Bugs validated (bugfixes)
- [ ] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### 🎉 GIF

_Please provide a suitable gif that describes your work on this pull request_

### CODE REVIEW

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/plugin/factory/StreamOfflinePluginFactory.kt [hunk 1]
```diff
@@ -71,15 +71,21 @@ public class StreamOfflinePluginFactory(
         }
     }
 
+    private var customRepositoryFactory: RepositoryFactory? = null
+
+    public fun setCustomRepositoryFactory(repositoryFactory: RepositoryFactory) {
```
Comment (JcMinarro) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/plugin/factory/StreamOfflinePluginFactory.kt:76]: What about injecting a `RepositoryFactory.Provider` on the constructor?
Our customers could need the currentUser to build their Repositories, and we should provide the userScope as well, then they can cancel their jobs when it is not needed anymore.

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/models/Filters.kt [hunk 2]
```diff
@@ -67,12 +67,21 @@ public object Filters {
...
     public fun and(vararg filters: FilterObject): FilterObject = AndFilterObject(filters.toSet())
 
+    @JvmStatic
+    public fun and(filters: List<FilterObject>): FilterObject = AndFilterObject(filters.toSet())
+
     @JvmStatic
     public fun or(vararg filters: FilterObject): FilterObject = OrFilterObject(filters.toSet())
 
+    @JvmStatic
+    public fun or(filters: List<FilterObject>): FilterObject = OrFilterObject(filters.toSet())
```
Comment (JcMinarro) [inline stream-chat-android-client/src/main/java/io/getstream/chat/android/client/models/Filters.kt:77]: Do we really need all those new methods?
If anyone wants to use our methods with a list, they can always use the [spread operator](https://kotlinlang.org/docs/functions.html#variable-number-of-arguments-varargs)

```
val listOfFilters = listOf<FilterObject>(...)
val orFilter = or(*listOfFliters.toTypedArray())
```

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/plugin/factory/StreamOfflinePluginFactory.kt (hunk 1 above)
Comment (leandroBorgesFerreira) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/plugin/factory/StreamOfflinePluginFactory.kt:76]: I didn't notice the `ChatClient.withRepositoryFactoryProvider` method.

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/api/models/querysort/QuerySorter.kt [hunk 3]
```diff
@@ -16,12 +16,16 @@
...
 
+import io.getstream.chat.android.client.api.models.querysort.internal.SortSpecification
+
 /**
  * Interface for the sorter of the SDK. Its implementations must provide a comparator to be used for
  * sorting collections of data in the SDK.
  */
 public interface QuerySorter<T : Any> {
 
+    public var sortSpecifications: List<SortSpecification<T>>
```
Comment (adasiewiczr) [inline stream-chat-android-client/src/main/java/io/getstream/chat/android/client/api/models/querysort/QuerySorter.kt:27]: Would be nice to add KDoc here

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/api/models/querysort/SortDirection.kt [hunk 4]
```diff
@@ -22,5 +22,14 @@ public enum class SortDirection(public val value: Int) {
     DESC(-1),
 
     /** Ascending sort order. */
-    ASC(1)
+    ASC(1);
+
+    public companion object {
+        public fun fromNumber(value: Int): SortDirection =
```
Comment (adasiewiczr) [inline stream-chat-android-client/src/main/java/io/getstream/chat/android/client/api/models/querysort/SortDirection.kt:28]: and here as well

**Review by adasiewiczr — APPROVED** [review]

**Pull request merged**